File: backend-stack-c/1_EA-and-backfill-worker-on-contabo-vps/v2_29_data_pipeline_architecture/calculation-split-between-mt5-and-python-PENDING-PROJECT/zscore_candle.py
"""Z-Score OHLC Candle calculation — literal Python port of
mql5-indicators/mql5-indicator-export-selection/USE/mq5/zscore-ohlc-candle-export.mq5

Port fidelity notes (MQL5 source references):
- Body size        = MathAbs(close - open)                     (line ~429)
- Z-score          = rolling window of InpZScoreLength bars,
                     SAMPLE variance (n - 1 divisor)           (CalculateZScore)
- Classification   = signed z-score vs thresholds; bullish if
                     close >= open                             (ClassifyCandle)
- body_direction   = 1 / -1 / 0 (close > / < / == open)        (export section)
- Exported "body_size" column = MathAbs(z-score), NOT the raw
  body size                                                    (export section)

Classification codes (= MQL5 color-index enum order):
  0 UP_NORMAL, 1 UP_LARGE, 2 UP_EXTREME, 3 DOWN_NORMAL, 4 DOWN_LARGE, 5 DOWN_EXTREME
"""

import math
from dataclasses import dataclass
from typing import List, Optional


@dataclass
class ZScoreCandleParams:
    """User-configurable parameters (MQL5 inputs)."""
    zscore_length: int = 432    # InpZScoreLength
    threshold_z1: float = 1.5   # InpThresholdZ1 (Large)
    threshold_z2: float = 2.5   # InpThresholdZ2 (Extreme)


@dataclass
class ZScoreCandleResult:
    body_direction: int             # 1 bullish, -1 bearish, 0 doji
    body_size_raw: float            # |close - open| (internal buffer value)
    zscore: Optional[float]         # signed z-score; None during warm-up
    body_size_export: Optional[float]  # |z-score| — the exported "body_size" column
    classification: Optional[int]   # 0-5; None during warm-up


def _zscore_at(body_sizes: List[float], position: int, length: int) -> Optional[float]:
    """CalculateZScore(): rolling sample z-score over `length` bars ending at position."""
    if position - (length - 1) < 0:
        return None  # warm-up: MQL5 starts at start index >= length
    s = 0.0
    s2 = 0.0
    for j in range(length):
        v = body_sizes[position - j]
        s += v
        s2 += v * v
    mean = s / length
    variance = (s2 - s * mean) / (length - 1)
    std_dev = math.sqrt(variance) if variance > 0 else 0.0
    return (body_sizes[position] - mean) / std_dev if std_dev != 0 else 0.0


def _classify(zscore: float, is_bullish: bool, p: ZScoreCandleParams) -> int:
    """ClassifyCandle(): signed z-score vs thresholds (bullish = close >= open)."""
    if is_bullish:
        if zscore >= p.threshold_z2:
            return 2  # UP_EXTREME
        if zscore >= p.threshold_z1:
            return 1  # UP_LARGE
        return 0      # UP_NORMAL
    if zscore >= p.threshold_z2:
        return 5      # DOWN_EXTREME
    if zscore >= p.threshold_z1:
        return 4      # DOWN_LARGE
    return 3          # DOWN_NORMAL


def calculate(opens: List[float], closes: List[float],
              params: Optional[ZScoreCandleParams] = None) -> List[ZScoreCandleResult]:
    """Compute the full z-score candle series (oldest-first arrays, MQL5 buffer order).

    Bars earlier than `zscore_length - 1` are warm-up: direction/raw body size are
    produced, z-score/classification are None (MQL5 leaves those buffers empty).
    """
    p = params or ZScoreCandleParams()
    body_sizes = [abs(c - o) for o, c in zip(opens, closes)]
    results: List[ZScoreCandleResult] = []

    for i, (o, c) in enumerate(zip(opens, closes)):
        direction = 1 if c > o else (-1 if c < o else 0)
        z = _zscore_at(body_sizes, i, p.zscore_length)
        if z is None:
            results.append(ZScoreCandleResult(direction, body_sizes[i], None, None, None))
        else:
            cls = _classify(z, c >= o, p)
            results.append(ZScoreCandleResult(direction, body_sizes[i], z, abs(z), cls))
    return results

File: lib/currency-index-comparison/zscore-candle.ts
/**
 * Currency Index Comparison (PRO) -- Z-score candles ("MC") on index candles.
 *
 * Pure and client-safe, so the page recomputes instantly when a slider moves.
 *
 * Ground truth is zscoreohlccandleexport_v2_29.mq5 (backend-stack-c/.../mq5/),
 * the indicator zscore_candle.py ports and golden_certification.py certified
 * (M5 and M15 PASS). Ported from the MQL5 rather than the Python, which differ
 * in one place: OnCalculate() starts at bar `InpZScoreLength`, so the first
 * classified bar is index `length` and bar 0 never enters a window, while
 * zscore_candle.py classifies from index `length - 1`. The certification
 * compared the last 500 bars only, so it never reached that edge.
 *
 * - Body size = |close - open| (CalculateBodySizeStats).
 * - Z-score = (body - mean) / sample stddev over the `length` bars ending at
 *   the bar, itself included (CalculateZScore). A zero stddev gives 0. A
 *   slightly negative float variance gives NaN in MQL5's MathSqrt() and in
 *   Math.sqrt() alike, and a NaN z-score classifies as Normal in both.
 * - Class: bullish when close >= open; Large at z >= threshold 1, Extreme at
 *   z >= threshold 2 (ClassifyCandle). Extreme is tested first, so a second
 *   threshold below the first simply leaves Large empty, as in MQL5.
 *
 * The MQL5 draws Normal candles in clrNONE; this page draws only Large and
 * Extreme candles, per Davin's spec.
 *
 * @module lib/currency-index-comparison/zscore-candle
 */

import type { IndexCandle } from './series';

/** The page's default `InpZScoreLength` (Davin's spec; the .mq5 input is 432). */
export const DEFAULT_ZSCORE_LENGTH = 54;
/** `InpThresholdZ1`. */
export const DEFAULT_ZSCORE_THRESHOLD_1 = 1.5;
/** `InpThresholdZ2`. */
export const DEFAULT_ZSCORE_THRESHOLD_2 = 2.5;

/** The sample-variance divisor is `length - 1`, so 2 is the smallest usable. */
export const ZSCORE_LENGTH_MIN = 2;
export const ZSCORE_LENGTH_MAX = 500;
export const ZSCORE_THRESHOLD_MIN = 0.1;
export const ZSCORE_THRESHOLD_MAX = 5;

/** The .mq5's CANDLE_TYPE enum, in its order. */
export const ZSCORE_CANDLE_CLASSES = [
  'up-normal',
  'up-large',
  'up-extreme',
  'down-normal',
  'down-large',
  'down-extreme',
] as const;

export type ZScoreCandleClass = (typeof ZSCORE_CANDLE_CLASSES)[number];

/** The four classes the page draws; Normal candles stay invisible. */
export type HighlightedZScoreClass = Exclude<
  ZScoreCandleClass,
  'up-normal' | 'down-normal'
>;

export interface ZScoreCandleParams {
  length: number;
  thresholdZ1: number;
  thresholdZ2: number;
}

export const DEFAULT_ZSCORE_PARAMS: ZScoreCandleParams = {
  length: DEFAULT_ZSCORE_LENGTH,
  thresholdZ1: DEFAULT_ZSCORE_THRESHOLD_1,
  thresholdZ2: DEFAULT_ZSCORE_THRESHOLD_2,
};

/** Candles needed before the first bar is classified: `length + 1`. */
export function zscoreBarsNeeded(length: number): number {
  return length + 1;
}

/**
 * One class per candle, null during warm-up. All null when there are fewer
 * than `length` candles (`if(rates_total < InpZScoreLength) return 0;`).
 */
export function zscoreCandleClasses(
  candles: readonly IndexCandle[],
  params: ZScoreCandleParams = DEFAULT_ZSCORE_PARAMS
): (ZScoreCandleClass | null)[] {
  const { length, thresholdZ1, thresholdZ2 } = params;
  const n = candles.length;
  const result = new Array<ZScoreCandleClass | null>(n).fill(null);
  if (length < ZSCORE_LENGTH_MIN || n < length) return result;

  const body = candles.map((c) => Math.abs(c.close - c.open));

  for (let i = length; i < n; i++) {
    let sum = 0;
    let sum2 = 0;
    for (let j = 0; j < length; j++) {
      const v = body[i - j] as number;
      sum += v;
      sum2 += v * v;
    }
    const mean = sum / length;
    const variance = (sum2 - sum * mean) / (length - 1);
    const stdDev = Math.sqrt(variance);
    const zScore = stdDev !== 0 ? ((body[i] as number) - mean) / stdDev : 0;

    const candle = candles[i] as IndexCandle;
    const bullish = candle.close >= candle.open;
    let code: number;
    if (zScore >= thresholdZ2) code = bullish ? 2 : 5;
    else if (zScore >= thresholdZ1) code = bullish ? 1 : 4;
    else code = bullish ? 0 : 3;
    result[i] = ZSCORE_CANDLE_CLASSES[code] as ZScoreCandleClass;
  }
  return result;
}

export function isHighlightedZScoreClass(
  cls: ZScoreCandleClass | null
): cls is HighlightedZScoreClass {
  return cls !== null && cls !== 'up-normal' && cls !== 'down-normal';
}
